### ISSUE TITLE
Can not access Keyvault from ASP.NET Core Web API using Pod Identity

### ISSUE THREAD

**Issue by pantanuj** [Azure/aad-pod-identity#233]
I have deployed a aspnet core web api and created pod identity and bindings. However, the api does not return a result and the error in pod log states 
Application startup exception: Microsoft.Azure.Services.AppAuthentication.AzureServiceTokenProviderException: Parameters: Connection String: [No connection string specified], Resource: https://vault.azure.net, Authority: https://login.windows.net/bae7949e-394f-4f3f-85a7-ef866f2e2450. Exception Message: Tried the following 3 methods to get an access token, but none of them worked.
Parameters: Connection String: [No connection string specified], Resource: https://vault.azure.net, Authority: https://login.windows.net/bae7949e-394f-4f3f-85a7-ef866f2e2450. Exception Message: Tried to get token using Managed Service Identity. Access token could not be acquired. MSI ResponseCode: Forbidden, Response: adal: Refresh request failed. Status Code = '400'. Response body: {"error":"invalid_request","error_description":"Identity not found"}

Parameters: Connection String: [No connection string specified], Resource: https://vault.azure.net, Authority: https://login.windows.net/bae7949e-394f-4f3f-85a7-ef866f2e2450. Exception Message: Tried to get token using Visual Studio. Access token could not be acquired. Environment variable LOCALAPPDATA not set.
Parameters: Connection String: [No connection string specified], Resource: https://vault.azure.net, Authority: https://login.windows.net/bae7949e-394f-4f3f-85a7-ef866f2e2450. Exception Message: Tried to get token using Azure CLI. Access token could not be acquired. /bin/bash: az: No such file or directory

Pod Identity

```
apiVersion: "aadpodidentity.k8s.io/v1"
kind: AzureIdentity
metadata:
 name: transactionapiidentity
spec:
 type: 0
 ResourceID: "/subscriptions/b31da45e-98ee-4988-ab63-451b2f178c90/resourcegroups/astra-sea-dev/providers/Microsoft.ManagedIdentity/userAssignedIdentities/LoyaltyPodIdentity"
 ClientID: "318e0f02-e0b0-41c4-9de7-6e5cfd6fc5ad"
```

Binding
```
apiVersion: "aadpodidentity.k8s.io/v1"
kind: AzureIdentityBinding
metadata:
 name: transactionapiidentity-binding
spec:
 AzureIdentity: transactionapiidentity
 Selector: transaction-api
```
Deployment
```
apiVersion: apps/v1
kind: Deployment
metadata:
 name: transactionapi-deployment 
spec:
 replicas: 3
 selector:
  matchLabels:
   app: transaction-api 
 template:
  metadata:
   labels:
    app: transaction-api
    aadpodidbinding: "transaction-api"
  spec:
   containers:
   - name: transaction-api 
     image: imagename:latest
     ports:
     - containerPort: 8080
   imagePullSecrets:
     - name: test
```
AzureServiceTokenProvider version 1.03

MIC Log
`I0606 09:40:33.600747       1 event.go:218] Event(v1.ObjectReference{Kind:"AzureIdentityBinding", Namespace:"default", Name:"transactionapiidentity-binding", UID:"ca77a82f-883e-11e9-81cb-56ca9633156d", APIVersion:"aadpodidentity.k8s.io/v1", ResourceVersion:"1819422", FieldPath:""}): type: 'Normal' reason: 'binding applied' Binding transactionapiidentity-binding applied on node aks-agentpool-25629171-1 for pod transactionapi-deployment-857978bd8d-sqw72-default-transactionapiidentity`

NMI Log
```
time="2019-06-06T11:24:53Z" level=info msg="node(aks-agentpool-25629171-1) hostip(10.240.0.66) metadataaddress(169.254.169.254:80) nmiport(2579)"
time="2019-06-06T11:24:53Z" level=info msg="Rules for table(nat) chain(aad-metadata) rules(-N aad-metadata, -A aad-metadata ! -s 127.0.0.1/32 -d 169.254.169.254/32 -p tcp -m tcp --dport 80 -j DNAT --to-destination 10.240.0.66:2579, -A aad-metadata -j RETURN)"
time="2019-06-06T11:25:53Z" level=info msg="node(aks-agentpool-25629171-1) hostip(10.240.0.66) metadataaddress(169.254.169.254:80) nmiport(2579)"
time="2019-06-06T11:25:53Z" level=info msg="Rules for table(nat) chain(aad-metadata) rules(-N aad-metadata, -A aad-metadata ! -s 127.0.0.1/32 -d 169.254.169.254/32 -p tcp -m tcp --dport 80 -j DNAT --to-destination 10.240.0.66:2579, -A aad-metadata -j RETURN)"
time="2019-06-06T11:26:53Z" level=info msg="node(aks-agentpool-25629171-1) hostip(10.240.0.66) metadataaddress(169.254.169.254:80) nmiport(2579)"
time="2019-06-06T11:26:53Z" level=info msg="Rules for table(nat) chain(aad-metadata) rules(-N aad-metadata, -A aad-metadata ! -s 127.0.0.1/32 -d 169.254.169.254/32 -p tcp -m tcp --dport 80 -j DNAT --to-destination 10.240.0.66:2579, -A aad-metadata -j RETURN)"
time="2019-06-06T11:27:53Z" level=info msg="node(aks-agentpool-25629171-1) hostip(10.240.0.66) metadataaddress(169.254.169.254:80) nmiport(2579)"
time="2019-06-06T11:27:53Z" level=info msg="Rules for table(nat) chain(aad-metadata) rules(-N aad-metadata, -A aad-metadata ! -s 127.0.0.1/32 -d 169.254.169.254/32 -p tcp -m tcp --dport 80 -j DNAT --to-destination 10.240.0.66:2579, -A aad-metadata -j RETURN)"
time="2019-06-06T11:28:53Z" level=info msg="node(aks-agentpool-25629171-1) hostip(10.240.0.66) metadataaddress(169.254.169.254:80) nmiport(2579)"
time="2019-06-06T11:28:53Z" level=info msg="Rules for table(nat) chain(aad-metadata) rules(-N aad-metadata, -A aad-metadata ! -s 127.0.0.1/32 -d 169.254.169.254/32 -p tcp -m tcp --dport 80 -j DNAT --to-destination 10.240.0.66:2579, -A aad-metadata -j RETURN)"
```

**Comment by samrobillard** [comment]
Hi,

It looks like you're missing the 'args' and 'env' parts of your deployment. Can you try to add that part under 'containers' in your yaml?

        args:
          - "--subscriptionid=<SubID>"
          - "--clientid=<transaction-api>"
          - "--resourcegroup=<ResourceGroup>"
          - "--aad-resourcename=https://vault.azure.net"
        env:
        - name: MY_POD_NAME
          valueFrom:
            fieldRef:
              fieldPath: metadata.name
        - name: MY_POD_NAMESPACE
          valueFrom:
            fieldRef:
              fieldPath: metadata.namespace
        - name: MY_POD_IP
          valueFrom:
            fieldRef:
              fieldPath: status.podIP

Cheers,

**Comment by aramase** [comment]
@pantanuj Can you post the result for `kubectl get azureassignedidentity`? The logs you posted for mic aren't sufficient to determine if there is any issue with aad-pod-identity. If you don't have any azureassignedidentities, you redeploy mic with arg `--v=6`. This will give us more verbose logs that can help us determine the reason for the issue.

**Comment by hermanho** [comment]
I have faced the same error in pod. However, the keyvault can be retrieved after the pod is restarted few times automatically. Here is the mic log file for your reference.

[logs-from-mic-in-mic-7ddbc68f8-srrsx.txt](https://github.com/Azure/aad-pod-identity/files/3299511/logs-from-mic-in-mic-7ddbc68f8-srrsx.txt)

**Comment by aramase** [comment]
@pantanuj We have released v1.5.2 (https://github.com/Azure/aad-pod-identity/releases/tag/v1.5.2). This release includes a lot of new features and improvements. Please try it out and let us know if your issues are resolved.

**Comment by kkmsft** [comment]
@pantanuj: Closing this issue. Please reopen if this issue still exists in v1.5.2.
